### PR TITLE
Springs Component is a Stereotype

### PR DESCRIPTION
**Pull request by BrentDouglas**
This extends the spring integration to reflect that component
annotations are stereotypes. It adds some additional scopes declared
by spring and some additional validation on names and scopes.

### CODE REVIEW

**Comment by gsmet** [comment]
@mkouba could you take a look at this one: it makes some changes to the Arc extension and to Arc
@geoand your opinion might be helpful here too

**Comment by geoand** [comment]
This looks cool, thanks for the PR @BrentDouglas !

However since it's rather large, I would like to understand what it bring to the table for the end user. Would you mind expanding on that in some detail @BrentDouglas ?

**Comment by BrentDouglas** [comment]
@geoand No worries, I was just investigating how much effort it would be to migrate some of our projects to quarkus.

Re the largeness, most of the changes here are to test classes. I'm not sure how to elaborate on it too much, this just makes the arc mapping closer to how spring beans work. All the spring projects I have worked on recently have used this though I can't really comment on how common it is anywhere else. Is the goal here to help people with spring projects migrate onto this? If so I think that having spring support here working more like spring would be beneficial for those users, and if that's not the goal then why have any spring support at all? I don't imagine that the idea is to have people start new projects on quarkus using spring, but I would think that if there is going to be any spring support then it should be as close to spring as it can be right?

Re the arc changes, there are only really two functional changes. It splits up one arc build item into two so that it can expose the jandex instance that the arc deployer has to the spring extension and secondly it injects a list of additional stereotype classes so that arc doesn't have to know anything about springs stuff. The rest are mostly from the formatter.

**Comment by geoand** [comment]
> @geoand No worries, I was just investigating how much effort it would be to migrate some of our projects to quarkus.
> 

That's awesome to know, we definitely want your feedback :)

> Re the largeness, most of the changes here are to test classes. I'm not sure how to elaborate on it too much, this just makes the arc mapping closer to how spring beans work. All the spring projects I have worked on recently have used this though I can't really comment on how common it is anywhere else. Is the goal here to help people with spring projects migrate onto this? If so I think that having spring support here working more like spring would be beneficial for those users, and if that's not the goal then why have any spring support at all? I don't imagine that the idea is to have people start new projects on quarkus using spring, but I would think that if there is going to be any spring support then it should be as close to spring as it can be right?

The idea is for people that are used to Spring DI annotations to be able to feel comfortable with Quarkus, at least with the DI part. Then as people gain confidence then can stray into other realms that Quarkus provides.

But the real question is following: Do we invest a significant amount of time in making the already included Spring things work absolutely like Spring, or do make available the most common cases and then invest time in bringing in more things that Spring devs are used thus lowering the general bar of entry?

> 
> Re the arc changes, there are only really two functional changes. It splits up one arc build item into two so that it can expose the jandex instance that the arc deployer has to the spring extension and secondly it injects a list of additional stereotype classes so that arc doesn't have to know anything about springs stuff. The rest are mostly from the formatter.

I will defer to @mkouba about the Arc stuff. If he's OK with the changes there, then I will on my part look into the Spring things in more detail (although at first the work does pass the eye test - :+1: )

**Comment by BrentDouglas** [comment]
> > @geoand No worries, I was just investigating how much effort it would be to migrate some of our projects to quarkus.
> 
> That's awesome to know, we definitely want your feedback :)

So far it's too early to tell much but this was the first thing I noticed that looked pretty simple to fix.

> > Re the largeness, most of the changes here are to test classes. I'm not sure how to elaborate on it too much, this just makes the arc mapping closer to how spring beans work. All the spring projects I have worked on recently have used this though I can't really comment on how common it is anywhere else. Is the goal here to help people with spring projects migrate onto this? If so I think that having spring support here working more like spring would be beneficial for those users, and if that's not the goal then why have any spring support at all? I don't imagine that the idea is to have people start new projects on quarkus using spring, but I would think that if there is going to be any spring support then it should be as close to spring as it can be right?
> 
> The idea is for people that are used to Spring DI annotations to be able to feel comfortable with Quarkus, at least with the DI part. Then as people gain confidence then can stray into other realms that Quarkus provides.
> 
> But the real question is following: Do we invest a significant amount of time in making the already included Spring things work absolutely like Spring, or do make available the most common cases and then invest time in bringing in more things that Spring devs are used thus lowering the general bar of entry?

I guess that's really up to you guys. If you would like I can split out the arc stuff and look at spring changes externally so it would not detract from the core effort.

> > Re the arc changes, there are only really two functional changes. It splits up one arc build item into two so that it can expose the jandex instance that the arc deployer has to the spring extension and secondly it injects a list of additional stereotype classes so that arc doesn't have to know anything about springs stuff. The rest are mostly from the formatter.
> 
> I will defer to @mkouba about the Arc stuff. If he's OK with the changes there, then I will on my part look into the Spring things in more detail (although at first the work does pass the eye test -  )

I'm not terribly happy with all the format changes happening here, is arc using a different formatter than the rest of the repo or was it merged in from some other repo or something without being reformatted? If this the correct format, I will split the actual changes out from the format stuff so that this commit looks less confusing.

**Comment by geoand** [comment]
@BrentDouglas I am not sure about the formatting changes to be honest... Did you follow [this](https://github.com/quarkusio/quarkus/blob/master/CONTRIBUTING.md#ide-config-and-code-style) section on setting the proper formatting settings? If not, please do so :)

Other than the formatting issues, let's see what @mkouba has to say about the Arc changes before we proceed any further.

**Comment by BrentDouglas** [comment]
@geoand Yeah, the format here is from IDEA with the eclipse plugin, plus boosting the star import limit to 9999.

**Comment by geoand** [comment]
@BrentDouglas then perhaps the Arc stuff was not formatted properly when the formatting was first applied, I'm not really sure...

**Comment by mkouba** [comment]
AFAIK ArC code wasn't formatted as it's still treated as an independent project and will likely move to a separate repo in the future.

I did not have time to review this PR yet. Spring integration is a lower priority right now.

**Comment by stuartwdouglas** [comment]
@mkouba should we just use the same formatting for ArC? I did not reformat the independent projects as some of them where existing projects that definitely should not have been reformatted, but maybe it makes sense to do it for ArC as it may be in the main repo for a while?

**Comment by mkouba** [comment]
@stuartwdouglas yes, it makes a lot of sense. I'm going to send a PR today.

**Comment by geoand** [comment]
@BrentDouglas would it be possible for you to rebase your PR onto master once the PR that @mkouba sends lands there? 
Not having the formatting noise will make things a lot easier to review

**Comment by mkouba** [comment]
I'll ping @BrentDouglas when the formatting PR is merged.

**Comment by stuartwdouglas** [comment]
@mkouba Looks like the formatting issues have been fixed, so it might be easier to merge this first.

**Comment by stuartwdouglas** [comment]
This needs a rebase

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Modifier.java
Comment (geoand) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Modifier.java]: I don't really think this class is needed. Basically the only use for it is in the tests (it's use in `BeanDeployment` can be replaced with vanilla `Modiier`).
If you absolutely need the class, then I suggest to place it in the tests only

File: extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java
Comment (geoand) [inline extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java]: :+1: 

Can you please update the comments to explain *why* the dependents needs to be after the dependencies so future readers of the code don't have to figure that out on the consumer side?

File: extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java
Comment (geoand) [inline extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java]: It looks like these can be package-private

**Comment by geoand** [comment]
> This needs a bit more work to make scopes work properly at the moment. I am still actively working on it.

Cool, I will an appropriate label then. Feel free to remove it when you are finished with the implementation.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Modifier.java
Comment (BrentDouglas) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Modifier.java]: Ok, I have moved it into the DI part and refactored it so that's the only place its needed

File: extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java
Comment (BrentDouglas) [reply extensions/spring-di/deployment/src/main/java/io/quarkus/spring/di/deployment/SpringDIProcessor.java]: Done

**Comment by BrentDouglas** [comment]
So scopes work mostly OK here now. I don't really understand why I need to register the session scope in the DI module, I would have assumed it would have been added by the registration thing in undertow but without manually registering it here it was not showing up in the tests. This doesn't seem like the right thing to be doing though, is there something better that should be happening?

File: extensions/spring-di/deployment/pom.xml
Comment (mkouba) [inline extensions/spring-di/deployment/pom.xml]: @BrentDouglas The session context is registered in the undertow extension but you need to depend on the deployment module to make it work -> `io.quarkus:quarkus-undertow`.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java]: I think that we should rather fail in the `BeanDeployment.findStereotypes()` method when multiple scopes are declared on a stereotype annotation. There is no need to wait until a bean attempts to use it.

**Comment by mkouba** [comment]
So first of all thanks for your contribution! But TBH I'm a bit lost and it would be really helpful if you could try to explain a little what changes are included. For example what is `StereotypeScopesBuildItem` for?

File: extensions/spring-di/deployment/pom.xml
Comment (BrentDouglas) [reply extensions/spring-di/deployment/pom.xml]: Ok, Updated.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
Comment (BrentDouglas) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java]: Done.

**Comment by BrentDouglas** [comment]
`StereotypeScopesBuildItem` was just uses to pass things between build steps within the DI plugin because there was a circular dep when I added the SessionScoped thing (now removed). I have removed it as it is no longer needed.

**Comment by geoand** [comment]
@BrentDouglas is this ready for another round of reviews?

**Comment by BrentDouglas** [comment]
@geoand Yes I think so. I updated the commit message to hopefully try and explain why its doing what its doing and think I have addressed the other points from @mkouba

**Comment by mkouba** [comment]
@BrentDouglas So now I think I finally understand the idea and I like it. I have few ideas how to improve the ArC integration a little, e.g. merge the `FullArchiveIndexBuildItem` into `BeanArchiveBuildItem`. But I'm going to merge this and propose improvements in another PR (I'll ping you for review ;-).

**Review by mkouba — APPROVED** [review]

**Comment by geoand** [comment]
Thanks @BrentDouglas !

**Comment by mkouba** [comment]
@BrentDouglas see also https://github.com/quarkusio/quarkus/pull/1485

**Pull request merged**
